src/System.Text.JsonLab/System/Text/Json/Utf8JsonReaderStream.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.IO;
+
+namespace System.Text.JsonLab
+{
+    public ref struct Utf8JsonReaderStream
+    {
+        private Utf8JsonReader _jsonReader;
+        private Span<byte> _span;
+        private Stream _stream;
+        private byte[] _buffer;
+        private bool _isFinalBlock;
+        private long _consumed;
+
+        const int FirstSegmentSize = 1_024; //TODO: Is this necessary?
+        const int StreamSegmentSize = 4_096;
+
+        public JsonTokenType TokenType => _jsonReader.TokenType;
+        public ReadOnlySpan<byte> Value => _jsonReader.Value;
+        public long Consumed => _consumed + _jsonReader.Consumed;
+
+        public Utf8JsonReaderStream(Stream jsonStream)
+        {
+            if (!jsonStream.CanRead)
+                JsonThrowHelper.ThrowArgumentException("Stream must be readable");
+
+            _buffer = ArrayPool<byte>.Shared.Rent(FirstSegmentSize);
+            int numberOfBytes = jsonStream.Read(_buffer, 0, FirstSegmentSize);
+            _span = _buffer.AsSpan(0, numberOfBytes);
+            _stream = jsonStream;
+
+            _isFinalBlock = numberOfBytes < FirstSegmentSize;
+            _jsonReader = new Utf8JsonReader(_span, _isFinalBlock);
+            _consumed = 0;
+        }
+
+        public bool Read()
+        {
+            bool result = _jsonReader.Read();
+            if (!result && !_isFinalBlock)
+            {
+                return ReadNext();
+            }
+            return result;
+        }
+
+        private bool ReadNext()
+        {
+            bool result = false;
+
+            do
+            {
+                _consumed += _jsonReader.Consumed;
+                int leftOver = _span.Length - (int)_jsonReader.Consumed;
+                int amountToRead = StreamSegmentSize;
+                if (leftOver > 0)
+                {
+                    _stream.Position -= leftOver;
+
+                    // TODO: Should this be a settable property?
+                    if (leftOver >= 1_000_000)
+                    {
+                        // A single JSON token exceeds 1 MB in size . In such a rare case, allocate.
+                        byte[] maxBuffer = new byte[2_000_000_000];
```

Comment:
2GB?

Reply:
Yes, I realize this is odd behaviour. I will fix this up. I didn't put much thought into this corner case since I want to understand what the expectation is here.

Reply:
I'm just not understanding why any value >= 1_000_000 would mean you should allocate a buffer up to 2000x times as large.

Reply:
K, will double over 1 MB instead.